### PR TITLE
WIP Configurable default formatters for RAD reports

### PR DESCRIPTION
**Pull request by holyjak**
Override :com.fulcrologic.rad.report/type->formatter to install
globally a custom formatter for a type.

### CODE REVIEW

File: src/main/com/fulcrologic/rad/report.cljc [hunk 1]
```diff
@@ -520,34 +519,21 @@
...
    returns the formatted value of that column using the field formatter(s) defined
    on the column attribute or report. If no formatter is provided a default formatter
    will be used."
-  [report-instance row-props {::keys      [field-formatter]
+  [report-instance row-props {::keys      [field-formatter type]
                               ::attr/keys [qualified-key] :as column-attribute}]
   (let [value                  (get row-props qualified-key)
         report-field-formatter (comp/component-options report-instance ::field-formatters qualified-key)
+        {::app/keys [runtime-atom]} (comp/any->app report-instance)
+        default-formatter      (some-> runtime-atom deref :com.fulcrologic.rad/controls ::type->formatter type)
```
Comment (awkay) [inline src/main/com/fulcrologic/rad/report.cljc:532]: So, if these are default formatters for a given type, why is the type in the ::report ns?  There is already a type on the attribute. The pattern is that forms/reports/controls have *styles*, and the data has *types.

**Comment by awkay** [comment]
See comment on other repo.  Default formatters belong int he RAD library itself. Platform UI plugins are about just that: support that is specific to a specific platform (native/web) and style (semantic UI vs whatever).

File: src/main/com/fulcrologic/rad/report.cljc (hunk 1 above)
Comment (holyjak) [reply src/main/com/fulcrologic/rad/report.cljc:532]: That was a mistake, the type should have been one line down, here 

```
::attr/keys [qualified-key type] :as column-attribute
```

**Comment by holyjak** [comment]
I have updated the PR. I agree fully what you wrote in the other PR.

However I am not sure whether it makes sense to use styles for formatters. They are used I believe in forms so that it is possible to override how 1) a particular attribute is displayed everywhere or 2) a particular attribute is displayed in a particular form. For formatters this is already handled by `::report/field-formatter` on the attribute / `::report/field-formatters` on the report instance. If we wanted to reuse `::attributes/field-style` also for reports then we should perhaps deprecate `::report/field-formatter` on attributes to avoid the duplication?

**Comment by holyjak** [comment]
I have thought some more about this. What could make sense is:

1. The defaul formatters are hardcoded in the report ns but can be overridden by setting `::report/type->formatter` on the runtime-atom / using `set-default-format-style!`. It is `type->style->formatter` with the `:default` style, similar to `:com.fulcrologic.rad.form/type->style->control`. (Or leave out style and simply assume default in this fallback.)
2. I add, as you suggested, `report/install-formatter! report/set-default-format-style!` and the mirror form versions
3. We deprecate `::report/field-formatter` on attributes; `::attributes/field-style` will be used instead. Thus the API surface will be smaller, getting rid of one keyword, even though setting a formatter becomes more work (style + install-formatter!)
4. We replace `::report/field-formatters` with `::report/field-styles` or replace `::form/field-styles` with simply `::rad/field-styles` and use it both in reports and forms. (pros: a single kwd used everywhere; cons: not consistent with the other options for a form/report that are in the respective namespaces)

Thus it will become little more work to install a formatter on an attribute / report but both reports and forms become more similar.

**Comment by awkay** [comment]
Getting closer.  Let's let this sit a little longer. I'm not sure your proposal is right yet. There's a balance to strike between too many options and not enough.  I'm not ready to collapse the idea of field formatter and input style together like that.

**Comment by holyjak** [comment]
I see, the `:privacy` case is a good example where merging the two would not be optimal.

> it might make sense to leverage ::attribute/style as a facet of the equation (i.e. type + style leads to the selection when there is no override)

Since you already can override a formatter on a per attribute / per report format, using the style would only provide an alternative way to do it, no? The only benefit would be that if you want to have both custom formatting and input control, you could achieve that by setting a single option, ie the `:style`, instead of setting style+formatter.

Perhaps we could merge this PR as is and then think more about it. We can always add a one more arity to `install-formatter!` (taking also `style`) and add the `set-default-format-style!` later.

**Comment by awkay** [comment]
Yeah, I think we're there. I'll probably touch it up a bit, but this looks good. Thanks.

**Pull request merged**